## Основы анализа данных в Python

### Необязательное домашнее задание №2

*Алла Тамбовцева*

В этом ДЗ вам предлагается поработать с данными из файла `stars.csv`, который содержит
характеристики звезд. Показатели в файле:

* `Temperature (K)` : температура звезды по Кельвину;
* `Luminosity(L/Lo)` : светимость звезды в единицах светимости Солнца;
* `Radius(R/Ro)` : радиус звезды по отношению к Солнцу;
* `Absolute magnitude(Mv)` : абсолютная звездная величина;
* `Star type` : тип звезды (0 — коричневый карлик, 1 — красный карлик, 2 — белый карлик, 3 —
звезда главной последовательности, 4 — супергигант, 5 — гипергигант); 
* `Star color` : цвет звезды;
* `Spectral Class` : спектральный класс звезды.

Загрузите данные из файла `stars.csv`, сохраните их в датафрейм и выполните следующие задания.

In [1]:
import pandas as pd

In [2]:
stars = pd.read_csv("stars.csv")

### Задача 1

Выведите описательные статистики для всех числовых столбцов датафрейма. Проинтерпретируйте результаты для столбца `Temperature (K)`. Выведите описательные статистики для всех текстовых столбцов датафрейма. Звезды какого цвета встречаются чаще?

In [3]:
stars.describe()

Temperature (K)  Luminosity(L/Lo)  Radius(R/Ro)  \
count       240.000000        240.000000    240.000000   
mean      10497.462500     107188.361635    237.157781   
std        9552.425037     179432.244940    517.155763   
min        1939.000000          0.000080      0.008400   
25%        3344.250000          0.000865      0.102750   
50%        5776.000000          0.070500      0.762500   
75%       15055.500000     198050.000000     42.750000   
max       40000.000000     849420.000000   1948.500000   

       Absolute magnitude(Mv)   Star type  
count              240.000000  240.000000  
mean                 4.382396    2.500000  
std                 10.532512    1.711394  
min                -11.920000    0.000000  
25%                 -6.232500    1.000000  
50%                  8.313000    2.500000  
75%                 13.697500    4.000000  
max                 20.060000    5.000000

> В столбце `Temperature (K)` 240 заполненных ячеек, средняя температура звезды примерно 10497 градусов по Кельвину, медианная – 5776 градусов по Кельвину. Скорее всего, в выборке есть нехарактерно большие значения. Так и есть – верхний квартиль 15055.5, а максимум – 40000, температура 75% звезд не более, чем 15055.5, но есть звезды с температурой почти в 3 раза выше этой. С нетипично маленькими значениями все не так очевидно, с одной стороны, минимум тоже почти в 3 раза ниже, чем нижний квартиль, но, с другой стороны, порядок чисел здесь не такой, как в окрестности максимума, поэтому эта разница уже не будет такой заметной. <br>
Можете дополнительно построить ящик с усами и убедиться: <br>
`stars["Temperature (K)"].plot.box();`

In [4]:
stars.describe(include = "object")

Star color Spectral Class
count         240            240
unique         19              7
top           Red              M
freq          112            111

> Звезды красного цвета встречаются чаще (мода `Red`).

### Задача 2

Выведите таблицу частот для типа звезды (`Star type`). Запишите, чему равняется мода этого показателя.

In [5]:
# нет одной моды, подвох
# мода 0, 1, 2, 3, 4, 5

stars["Star type"].value_counts()

0    40
1    40
2    40
3    40
4    40
5    40
Name: Star type, dtype: int64

### Задача 3

Выведите таблицу частот для цвета звезды (`Star color`), отсортировав цвета по алфавиту.

In [6]:
stars["Star color"].value_counts().sort_index()

Blue                   55
Blue                    1
Blue White             10
Blue white              3
Blue white              1
Blue-White              1
Blue-white             26
Orange                  2
Orange-Red              1
Pale yellow orange      1
Red                   112
White                   7
White-Yellow            1
Whitish                 2
Yellowish               1
Yellowish White         3
white                   3
yellow-white            8
yellowish               2
Name: Star color, dtype: int64

### Задача 4

Выберите строки, соответствующие звездам с абсолютной звездной величиной меньше 0, и сохраните их в датафрейм `low`.

In [7]:
low = stars[stars["Absolute magnitude(Mv)"] < 0]
print(low.shape) # размер для проверки

(102, 7)


### Задача 5

Выберите строки, соответствующие звездам спектрального класса `O` или `M`, и сохраните их в датафрейм `stars_sm`.

In [8]:
stars_sm = stars[(stars["Spectral Class"] == "O") | 
                 (stars["Spectral Class"] == "M")]
print(stars_sm.shape) # размер для проверки

(151, 7)


### Задача 6

Выберите строки, соответствующие супергигантам с радиусом по отношению к Солнцу не ниже 50, и сохраните их в датафрейм `super_s`.

In [9]:
super_s = stars[(stars["Star type"] == 4) & 
      (stars["Radius(R/Ro)"] >= 50)]
print(super_s.shape) # размер для проверки

(18, 7)


### Задача 7

Добавьте в датафрейм столбец `WhiteDwarf`, который содержит значения 0 и 1, считая, что 1 ставится в случае, если звезда является белым карликом.

In [10]:
stars["WhiteDwarf"] = (stars["Star type"] == 2).astype(int)
stars["WhiteDwarf"].value_counts()

0    200
1     40
Name: WhiteDwarf, dtype: int64

### Задача 8

Добавьте в датафрейм столбец `Color`, который содержит следующие значения:

* `Red`: если в цвете звезды встречается слово *Red*;
* `Blue`: если в цвете звезды встречается слово *Blue*; 
* `Other`: иначе.

Встречается – слово может входить в состав других, например, `Red-white`,  и `Red ` с пробелом должны относиться к категории `Red` (и с `Blue` та же история).

In [11]:
# пишем функцию для проверки
# делаем в x все буквы маленькими на всякий случай
# (хотя здесь эти цвета всегда с заглавной буквы)

def color_agg(x):
    if "red" in x.lower():
        c = "Red"
    elif "blue" in x.lower():
        c = "Blue"
    else:
        c = "Other"
    return c
    
# применяем ее через apply()
stars["Color"] = stars["Star color"].apply(color_agg)
stars["Color"].value_counts()

Red      113
Blue      97
Other     30
Name: Color, dtype: int64

In [12]:
# вариант с lambda-функцией в одну строку
# если очень хочется (elif нет, в else вкладываем  
# еще один if в круглых скобках)

stars["Color"] = stars["Star color"].apply(lambda x: "Red" if "red" in x.lower() else 
                          ("Blue" if "blue" in x.lower() else "Other"))
stars["Color"].value_counts()

Red      113
Blue      97
Other     30
Name: Color, dtype: int64

### Задача 9

Добавьте в датафрейм столбец `LogLum`, который содержит логарифмированные значения светимости звезды `Luminosity(L/Lo)` (натуральный логарифм).

In [13]:
# log() из math не умеет работать со списками/массивами/столбцами, 
# а вот log() из numpy умеет

import numpy as np
stars["LogLum"] = np.log(stars["Luminosity(L/Lo)"])

### Задача 10

Сгруппируйте строки по значениям столбца `WhiteDwarf` (создается в задаче 7) и выведите для каждой группы:

* полный набор описательных статистик для радиуса звезды;
* медианное и среднее значение радиуса звезды.

In [14]:
stars.groupby("WhiteDwarf")["Radius(R/Ro)"].describe()

count        mean         std     min       25%     50%  \
WhiteDwarf                                                            
0           200.0  284.587192  554.656448  0.0570  0.240000  5.7125   
1            40.0    0.010728    0.001725  0.0084  0.009305  0.0102   

                  75%       max  
WhiteDwarf                       
0           80.250000  1948.500  
1            0.012025     0.015

In [15]:
stars.groupby("WhiteDwarf")["Radius(R/Ro)"].agg(["median", "mean"])

median        mean
WhiteDwarf                    
0           5.7125  284.587192
1           0.0102    0.010728